File: Data Structures & Algorithms/generate-parentheses/submission-1.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        self.res = []
        self.subset = ""
        def dfs(start, close):
            if close > start or start > n:
                return 
            if len(self.subset) == 2*n and start == close:
                self.res.append(self.subset)
                return 
            self.subset += '('
            dfs(start + 1, close)
            self.subset = self.subset[:-1]
            self.subset += ')'
            dfs(start, close + 1)
            self.subset = self.subset[:-1]
        dfs( 0, 0)
        return self.res

File: Data Structures & Algorithms/count-paths/submission-1.cpp
class Solution {
public:
    vector<vector<int>> visited;
    int uniquePaths(int m, int n) {
        visited.resize(m, vector<int>(n, -1));
        return dfs(m, n, 0, 0);
    }
    int res = 0;
    int dfs(int m, int n, int i, int j){
        if( i == (m-1) && j == (n-1)){
            return 1;
        }
        if (i >= m || j >= n) return 0;
        if(visited[i][j] != -1){
            return visited[i][j];
        }
        return visited[i][j] = dfs(m, n, i, j + 1) + dfs(m, n, i + 1, j);
    }
};
